File: src/main/java/rars/settings/EditorThemeSettings.java
package rars.settings;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rars.venus.editors.TokenStyle;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

import static rars.util.Utils.getColorAsHexString;

public final class EditorThemeSettings extends SettingsBase {
    private static final @NotNull Logger LOGGER = LogManager.getLogger();
    /**
     * Top level theme settings prefix.
     */
    private static final String THEME_PREFIX = "Theme";

    // region Preferences keys
    private static final String BACKGROUND = "Background";
    private static final String FOREGROUND = "Foreground";
    private static final String LINE_HIGHLIGHT = "LineHighlight";
    private static final String CARET = "Caret";
    private static final String SELECTION = "Selection";
    private static final String STYLES = "Styles";
    private static final String BOLD = "Bold";
    private static final String ITALIC = "Italic";
    private static final String UNDERLINE = "Underline";
    public static @NotNull EditorThemeSettings EDITOR_THEME_SETTINGS = new EditorThemeSettings(SETTINGS_PREFERENCES);

    // endregion Preferences keys
    private final @NotNull Preferences preferences;
    /**
     * The current theme in memory. You can make changes to this theme and then
     * call {@link #commitChanges()} to save the changes to the preferences.
     * You can also call {@link #discardChanges()} to revert the changes to the
     * state present in the preferences.
     */
    public @NotNull SettingsTheme currentTheme;
    private @NotNull SettingsTheme backupTheme;

    private EditorThemeSettings(final @NotNull Preferences preferences) {
        this.preferences = preferences;
        this.currentTheme = loadThemeFromPreferences();
        this.backupTheme = this.currentTheme.clone();
    }

    // region Preferences prefix methods

    private static @NotNull String foregroundPrefix(final @NotNull TokenSettingKey key) {
        return THEME_PREFIX + STYLES + key.ordinal() + FOREGROUND;
    }

    private static @NotNull String backgroundPrefix(final @NotNull TokenSettingKey key) {
        return THEME_PREFIX + STYLES + key.ordinal() + BACKGROUND;
    }

    private static @NotNull String boldPrefix(final @NotNull TokenSettingKey key) {
        return THEME_PREFIX + STYLES + key.ordinal() + BOLD;
    }

    private static @NotNull String italicPrefix(final @NotNull TokenSettingKey key) {
        return THEME_PREFIX + STYLES + key.ordinal() + ITALIC;
    }

    private static @NotNull String underlinePrefix(final @NotNull TokenSettingKey key) {
        return THEME_PREFIX + STYLES + key.ordinal() + UNDERLINE;
    }

    // endregion Preferences prefix methods

    /**
     * Commits the theme currently in memory to the preferences. If the commit
     * fails, the theme in memory will be reverted to the previous state.
     */
    public void commitChanges() {
        writeThemeToPreferences(this.currentTheme);
        try {
            this.preferences.flush();
            this.backupTheme = this.currentTheme;
            submit();
        } catch (final SecurityException se) {
            LOGGER.error("Unable to write to persistent storage for security reasons. Reverting to previous settings.");
            // The reason why we need to write the backup theme to the preferences
            // is because the Preferences API implementations are free to flush
            // the changes to disk at any time.
            writeThemeToPreferences(this.backupTheme);
            this.currentTheme = this.backupTheme;
        } catch (final BackingStoreException bse) {
            LOGGER.error("Unable to communicate with persistent storage. Reverting to previous settings.");
            writeThemeToPreferences(this.backupTheme);
            this.currentTheme = this.backupTheme;
        }
    }

    /**
     * Restores the state of the theme settings in memory to the state in the
     * preferences.
     */
    public void discardChanges() {
        this.currentTheme = this.backupTheme;
    }

    // region Preference writing methods
    private void writeThemeToPreferences(final @NotNull SettingsTheme settingsTheme) {
        this.preferences.put(THEME_PREFIX + BACKGROUND, getColorAsHexString(settingsTheme.backgroundColor));
        this.preferences.put(THEME_PREFIX + FOREGROUND, getColorAsHexString(settingsTheme.foregroundColor));
        this.preferences.put(THEME_PREFIX + LINE_HIGHLIGHT, getColorAsHexString(settingsTheme.lineHighlightColor));
        this.preferences.put(THEME_PREFIX + CARET, getColorAsHexString(settingsTheme.caretColor));
        this.preferences.put(THEME_PREFIX + SELECTION, getColorAsHexString(settingsTheme.selectionColor));
        settingsTheme.tokenStyles.forEach(this::writeTokenStyleToPreferences);
    }

    private void writeTokenStyleToPreferences(final @NotNull TokenSettingKey key, final @NotNull TokenStyle style) {
        putNullableColor(foregroundPrefix(key), style.foreground());
        putNullableColor(backgroundPrefix(key), style.background());
        this.preferences.putBoolean(boldPrefix(key), style.isBold());
        this.preferences.putBoolean(italicPrefix(key), style.isItalic());
        this.preferences.putBoolean(underlinePrefix(key), style.isUnderline());
    }

    /**
     * If the value is null, save "null" to the preferences.
     * Otherwise, save the value to the preferences.
     */
    private void putNullableColor(final @NotNull String key, final @Nullable Color value) {
        final @NotNull String valueString;
        if (value == null) {
            valueString = "null";
        } else {
            valueString = getColorAsHexString(value);
        }
        this.preferences.put(key, valueString);
    }

    // endregion Preference writing methods

    // region Preference loading methods

    private @NotNull SettingsTheme loadThemeFromPreferences() {
        final var defaultTheme = SettingsTheme.DEFAULT_THEME;
        final var background = loadColorFromPreferences(THEME_PREFIX + BACKGROUND, defaultTheme.backgroundColor);
        final var foreground = loadColorFromPreferences(THEME_PREFIX + FOREGROUND, defaultTheme.foregroundColor);
        final var lineHighlight = loadColorFromPreferences(THEME_PREFIX + LINE_HIGHLIGHT,
            defaultTheme.lineHighlightColor);
        final var caret = loadColorFromPreferences(THEME_PREFIX + CARET, defaultTheme.caretColor);
        final var selection = loadColorFromPreferences(THEME_PREFIX + SELECTION, defaultTheme.selectionColor);
        final var tokenStyles = loadTokenStylesFromPreferences(defaultTheme.tokenStyles);
        return new SettingsTheme(background, foreground, lineHighlight, caret, selection, tokenStyles);
    }

    private @NotNull Map<@NotNull TokenSettingKey, @NotNull TokenStyle> loadTokenStylesFromPreferences(
        final @NotNull Map<@NotNull TokenSettingKey, @NotNull TokenStyle> defaultColorScheme
    ) {
        final var styleMap = new HashMap<@NotNull TokenSettingKey, @NotNull TokenStyle>();
        for (final var type : TokenSettingKey.values()) {
            styleMap.put(type, loadTokenStyleFromPreferences(type, defaultColorScheme.get(type)));
        }
        return styleMap;
    }

    private @NotNull TokenStyle loadTokenStyleFromPreferences(final @NotNull TokenSettingKey key,
                                                              final @NotNull TokenStyle defaultStyle) {
        final var foreground = loadNullableColorFromPreferences(foregroundPrefix(key), defaultStyle.foreground());
        final var background = loadNullableColorFromPreferences(backgroundPrefix(key), defaultStyle.background());
        final var isBold = this.preferences.getBoolean(boldPrefix(key), defaultStyle.isBold());
        final var isItalic = this.preferences.getBoolean(italicPrefix(key), defaultStyle.isItalic());
        final var isUnderline = this.preferences.getBoolean(underlinePrefix(key), defaultStyle.isUnderline());
        return new TokenStyle(foreground, background, isBold, isItalic, isUnderline);
    }

    private @NotNull Color loadColorFromPreferences(final @NotNull String key, final @NotNull Color defaultValue) {
        final var value = this.preferences.get(key, null);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Color.decode(value);
        } catch (final NumberFormatException nfe) {
            LOGGER.error("Unable to decode color from preferences", nfe);
            return defaultValue;
        }
    }

    private @Nullable Color loadNullableColorFromPreferences(final @NotNull String key,
                                                             final @Nullable Color defaultValue) {
        final var value = this.preferences.get(key, null);
        if (value == null) {
            return defaultValue;
        } else if (value.equals("null")) {
            return null;
        }
        try {
            return Color.decode(value);
        } catch (final NumberFormatException nfe) {
            LOGGER.error("Unable to decode color from preferences", nfe);
            return defaultValue;
        }
    }

    // endregion Preference loading methods
}

File: src/main/java/rars/settings/EditorThemeSettings.kt
package rars.settings

import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import rars.util.ListenerDispatcher
import rars.venus.editors.TokenStyle
import java.util.prefs.BackingStoreException
import java.util.prefs.Preferences

interface EditorThemeSettings {
    val currentTheme: SettingsTheme
}

class EditorThemeSettingsImpl(private val preferences: Preferences) : EditorThemeSettings {
    private val onChangeDispatcher = ListenerDispatcher<Void?>()

    @JvmField
    val onChangeListenerHook = onChangeDispatcher.hook

    /**
     * The current theme in memory. You can make changes to this theme and then
     * call [saveSettingsToPreferences] to save the changes to the preferences.
     */
    private var _currentTheme: SettingsTheme = loadThemeFromPreferences()
    private var backupTheme: SettingsTheme = _currentTheme.clone()

    override var currentTheme: SettingsTheme
        get() = _currentTheme.clone()
        set(value) {
            this._currentTheme = value
        }

    /**
     * Commits the theme currently in memory to the preferences. If the commit
     * fails, the theme in memory will be reverted to the previous state.
     */
    fun saveSettingsToPreferences() {
        writeThemeToPreferences(this._currentTheme)
        try {
            this.preferences.flush()
            this.backupTheme = this._currentTheme
            this.onChangeDispatcher.dispatch(null)
        } catch (_: SecurityException) {
            LOGGER.error("Unable to write to persistent storage for security reasons. Reverting to previous settings.")
            // The reason why we need to write the backup theme to the preferences
            // is because the Preferences API implementations are free to flush
            // the changes to disk at any time.
            writeThemeToPreferences(this.backupTheme)
            this._currentTheme = this.backupTheme
        } catch (_: BackingStoreException) {
            LOGGER.error("Unable to communicate with persistent storage. Reverting to previous settings.")
            writeThemeToPreferences(this.backupTheme)
            this._currentTheme = this.backupTheme
        }
    }

    private fun writeThemeToPreferences(settingsTheme: SettingsTheme) {
        this.preferences.putColor(THEME_PREFIX + BACKGROUND, settingsTheme.backgroundColor)
        this.preferences.putColor(THEME_PREFIX + FOREGROUND, settingsTheme.foregroundColor)
        this.preferences.putColor(THEME_PREFIX + LINE_HIGHLIGHT, settingsTheme.lineHighlightColor)
        this.preferences.putColor(THEME_PREFIX + CARET, settingsTheme.caretColor)
        this.preferences.putColor(THEME_PREFIX + SELECTION, settingsTheme.selectionColor)
        settingsTheme.tokenStyles.forEach { (key, style) ->
            this.preferences.putTokenStyle(
                key.ordinal.toString(),
                style,
                THEME_PREFIX + STYLES
            )
        }
    }

    // region Preference loading methods
    private fun loadThemeFromPreferences(): SettingsTheme {
        val defaultTheme = SettingsTheme.DEFAULT_THEME
        val background = preferences.getColor(THEME_PREFIX + BACKGROUND, defaultTheme.backgroundColor, LOGGER)
        val foreground = preferences.getColor(THEME_PREFIX + FOREGROUND, defaultTheme.foregroundColor, LOGGER)
        val lineHighlight = preferences.getColor(THEME_PREFIX + LINE_HIGHLIGHT, defaultTheme.lineHighlightColor, LOGGER)
        val caret = preferences.getColor(THEME_PREFIX + CARET, defaultTheme.caretColor, LOGGER)
        val selection = preferences.getColor(THEME_PREFIX + SELECTION, defaultTheme.selectionColor, LOGGER)
        val tokenStyles = loadTokenStylesFromPreferences(defaultTheme.tokenStyles)
        return SettingsTheme(background, foreground, lineHighlight, caret, selection, tokenStyles)
    }

    private fun loadTokenStylesFromPreferences(
        defaultColorScheme: MutableMap<TokenSettingKey, TokenStyle>
    ): MutableMap<TokenSettingKey, TokenStyle> {
        val styleMap = HashMap<TokenSettingKey, TokenStyle>()
        for (type in TokenSettingKey.entries) {
            styleMap.put(
                type, this.preferences.getTokenStyle(
                    type.ordinal.toString(),
                    defaultColorScheme[type]!!,
                    THEME_PREFIX + STYLES,
                    LOGGER
                )
            )
        }
        return styleMap
    }
    // endregion Preference loading methods

    companion object {
        private val LOGGER: Logger = LogManager.getLogger(EditorThemeSettingsImpl::class.java)

        /**
         * Top level theme settings prefix.
         */
        private const val THEME_PREFIX = "Theme"

        // region Preferences keys
        private const val BACKGROUND = "Background"
        private const val FOREGROUND = "Foreground"
        private const val LINE_HIGHLIGHT = "LineHighlight"
        private const val CARET = "Caret"
        private const val SELECTION = "Selection"
        private const val STYLES = "Styles"
        private const val BOLD = "Bold"
        private const val ITALIC = "Italic"
        private const val UNDERLINE = "Underline"
        // endregion Preferences keys

        // region Preferences prefix methods
        private fun foregroundPrefix(key: TokenSettingKey): String = "$THEME_PREFIX$STYLES${key.ordinal}$FOREGROUND"
        private fun backgroundPrefix(key: TokenSettingKey): String = "$THEME_PREFIX$STYLES${key.ordinal}$BACKGROUND"
        private fun boldPrefix(key: TokenSettingKey): String = "$THEME_PREFIX$STYLES${key.ordinal}$BOLD"
        private fun italicPrefix(key: TokenSettingKey): String = "$THEME_PREFIX$STYLES${key.ordinal}$ITALIC"
        private fun underlinePrefix(key: TokenSettingKey): String = "$THEME_PREFIX$STYLES${key.ordinal}$UNDERLINE"
        // endregion Preferences prefix methods
    }
}